ecl/hqlcpp/hqlhtcpp.cpp

```diff
@@ -14033,12 +14033,26 @@ ABoundActivity * HqlCppTranslator::doBuildActivityDedup(BuildCtx & ctx, IHqlExpr
 
     if (!useHash)
     {
-        if (info.compareAllRows)
-            doBuildBoolFunction(instance->classctx, "compareAll", info.compareAllRows);
-        if (!info.keepLeft)
-            doBuildBoolFunction(instance->classctx, "keepLeft", info.keepLeft);
+        // Only necessary to create getFlags() if different from base class getFlags() - i.e. different from default
+        if ( !(!info.compareAllRows && !info.keepBest && info.keepLeft) )
```

Comment:
The double -ve does look confusing.  This section should be executed if any flags are different from the default.  The default flags are:
compareAllRows=false
keepBest=false
keepLeft=true

So, the thinking was if ( NOT (list of default flags)), then execute section. I can see how confusion it looks.  I'll simplify.

The second question relates to the handling of keepLeft.  Given the above defaults, getFlags() in the base class returns DFKeepLeft.  It is only necessary to create a new derived getFlags() if keepLeft!=True OR compareAllRows!=False OR keepBest!=False (i.e. if any the flags are different from the defaults).  

It may make more sense with this chart:

compareAllRows | keepBest | keepLeft | flag value
--------------- | --------------- | --------------- | ---------------
false | false | true | use base class getFlags()
false | false | false | Return 0
false | true | false | Return DFkeepbest
false | true | true | Return DFkeepbest \| DFkeepleft
true | false | true | Return DFcompareall \| DFkeepleft
true | false | false | Return DFcompareall
true | true | false | Return DFcompareall \| DFkeepbest
true | true | true | Return DFcompareall \| DFkeepbest \| keepLeft

The code is necessary as it is because keepLeft=true is the default.  If any of the defaults change, then all the flags need to be regenerated.